Sources/ComposableArchitecture/SwiftUI/ForEachStore.swift

```diff
@@ -78,64 +149,69 @@ import SwiftUI
+  public typealias ID = StatesCollection.ID
+
+  let store: Store<StatesCollection, (ID, EachAction)>
+  @ObservedObject var viewStore: ViewStore<StatesCollection, (ID, EachAction)>
+  let content: (ID, Store<EachState, EachAction>) -> EachContent
 
   /// Initializes a structure that computes views on demand from a store on a collection of data and
   /// an identified action.
   ///
   /// - Parameters:
   ///   - store: A store on an identified array of data and an identified action.
   ///   - content: A function that can generate content given a store of an element.
-  public init<EachContent>(
-    _ store: Store<IdentifiedArray<ID, EachState>, (ID, EachAction)>,
+  public init(
+    _ store: Store<StatesCollection, (ID, EachAction)>,
     @ViewBuilder content: @escaping (Store<EachState, EachAction>) -> EachContent
-  )
-  where
-    Data == IdentifiedArray<ID, EachState>,
-    Content == WithViewStore<
-      OrderedSet<ID>, (ID, EachAction), ForEach<OrderedSet<ID>, ID, EachContent>
-    >
-  {
-    self.data = store.state.value
-    self.content = WithViewStore(
+  ) {
+    self.store = store
+    self.viewStore = ViewStore(
       store,
-      observe: { $0.ids },
-      removeDuplicates: areOrderedSetsDuplicates
-    ) { viewStore in
-      ForEach(viewStore.state, id: \.self) { id -> EachContent in
-        // NB: We cache elements here to avoid a potential crash where SwiftUI may re-evaluate
-        //     views for elements no longer in the collection.
-        //
-        // Feedback filed: https://gist.github.com/stephencelis/cdf85ae8dab437adc998fb0204ed9a6b
-        var element = store.state.value[id: id]!
-        return content(
-          store.scope(
-            state: {
-              element = $0[id: id] ?? element
-              return element
-            },
-            action: { (id, $0) }
-          )
-        )
-      }
-    }
+      observe: { $0 },
+      removeDuplicates: { $0.areDuplicateIDs(other: $1) }
+    )
+    self.content = { content($1) }
+  }
+
+  /// Initializes a structure that computes views on demand from a store on a collection of data and
+  /// an identified action.
+  ///
+  /// - Parameters:
+  ///   - store: A store on an identified array of data and an identified action.
+  ///   - content: A function that can generate content given a store of an element and its
+  ///   corresponding identifier.
+  public init(
+    _ store: Store<StatesCollection, (ID, EachAction)>,
+    @ViewBuilder content: @escaping (ID, Store<EachState, EachAction>) -> EachContent
+  ) {
+    self.store = store
+    self.viewStore = ViewStore(
+      store,
+      observe: { $0 },
+      removeDuplicates: { $0.areDuplicateIDs(other: $1) }
+    )
+    self.content = content
   }
 
   public var body: some View {
-    self.content
+    ForEach(viewStore.stateIDs, id: \.self) { stateID in
+      // TODO: Message if `nil`?
+      let state = viewStore.state.extract(tag: stateID)!
```

Comment:
As we don't have the guarantee we had with `IdentifiedArray`, it is probably better to issue a runtime warning of the user's conformance to `IdentifiedStateCollection` is deficient.